File: Linear_Control_Course/Chapter30/Lesson2/Chapter30_Lesson2.m
% Parameters
alpha_values = [0.2, 0.9];
k = 2;

s = tf('s');

for idx = 1:numel(alpha_values)
    alpha = alpha_values(idx);

    % 2x2 coupled plant: G(s) = 1/(s+1)*[1 alpha; alpha 1]
    G11 = 1/(s + 1);
    G22 = 1/(s + 1);
    G12 = alpha/(s + 1);
    G21 = alpha/(s + 1);
    G = [G11 G12; G21 G22];

    % Decentralized proportional controller
    C = k * eye(2);

    % Loop transfer and closed-loop from r to y
    L = G * C;
    I2 = eye(2);
    T = feedback(L, I2);  % T(s) = L(I+L)^(-1)

    % Step in r1 only
    t = 0:0.01:10;
    r = [ones(size(t)); zeros(size(t))];
    [y, t_out] = lsim(T, r', t);

    figure;
    plot(t_out, y(:,1), 'b', 'LineWidth', 1.5); hold on;
    plot(t_out, y(:,2), 'r--', 'LineWidth', 1.5);
    grid on;
    xlabel('Time (s)');
    ylabel('Outputs');
    title(sprintf('Step in r1, alpha = %.2f', alpha));
    legend('y1', 'y2');
end

File: Tutorials/Codes/Linear_Control_Course/Chapter30/Lesson2/Chapter30_Lesson2.py
import numpy as np
import matplotlib.pyplot as plt
import control as ctl

# 2x2 coupled plant: G(s) = 1/(s+1) * [[1, alpha], [alpha, 1]]
s = ctl.TransferFunction.s

def make_plant(alpha):
    g11 = 1/(s + 1)
    g22 = 1/(s + 1)
    g12 = alpha/(s + 1)
    g21 = alpha/(s + 1)
    G = ctl.TransferFunction([[g11, g12],
                              [g21, g22]])
    return G

def closed_loop(alpha, k):
    G = make_plant(alpha)
    # Decentralized proportional controller: C(s) = k * I
    C = ctl.TransferFunction([[k, 0],
                              [0, k]])
    L = G * C
    # Closed-loop from r to y: T(s) = L (I + L)^(-1)
    I2 = ctl.tf2ss(ctl.TransferFunction([[1, 0],
                                         [0, 1]]))
    T = ctl.feedback(L, I2)  # feedback(L, I) = L (I+L)^(-1)
    return T

k = 2.0
alphas = [0.2, 0.9]
t = np.linspace(0, 10, 1000)

# Step in r1, r2 = 0
r1 = np.ones_like(t)
r2 = np.zeros_like(t)
R = np.vstack((r1, r2))

plt.figure()
for alpha in alphas:
    T = closed_loop(alpha, k)
    tout, y, _ = ctl.forced_response(T, T=t, U=R)
    plt.plot(tout, y[0, :], label=f"y1, alpha={alpha}")
    plt.plot(tout, y[1, :], "--", label=f"y2, alpha={alpha}")

plt.xlabel("Time (s)")
plt.ylabel("Outputs y1, y2")
plt.title("Step in r1 with decentralized control for different coupling alpha")
plt.legend()
plt.grid(True)
plt.show()
